pennylane/interfaces/batch/__init__.py

```diff
@@ -0,0 +1,124 @@
+            functions available in ``qml.gradients`` are only supported on the backward
+            pass.
+
+
+    Returns:
+        list[list[float]]: A nested list of tape results. Each element in
+        the returned list corresponds in order to the provided tapes.
+
+    **Example**
+
+    Consider the following cost function:
+
+    .. code-block:: python
+
+        dev = qml.device("lightning.qubit", wires=2)
+
+        def cost_fn(params, x):
+            with qml.tape.JacobianTape() as tape1:
+                qml.RX(params[0], wires=0)
+                qml.RY(params[1], wires=0)
+                qml.expval(qml.PauliZ(0))
+
+            with qml.tape.JacobianTape() as tape2:
+                qml.RX(params[2], wires=0)
+                qml.RY(x[0], wires=1)
+                qml.CNOT(wires=[0, 1])
+                qml.probs(wires=0)
+
+            tapes = [tape1, tape2]
+
+            # execute both tapes in a batch on the given device
+            res = execute(tapes, dev, qml.gradients.param_shift)
+
+            return res[0][0] + res[1][0, 0] - res[1][0, 1]
+
+    In this cost function, two **independent** quantum tapes are being
+    constructed; one returning an expectation value, the other probabilities.
+    We then batch execute the two tapes, and reduce the results to obtain
+    a scalar.
+
+    Let's execute this cost function while tracking the gradient:
+
+    >>> params = np.array([0.1, 0.2, 0.3], requires_grad=True)
+    >>> x = np.array([0.5], requires_grad=True)
+    >>> cost_fn(params, x)
+    1.9305068163274222
+
+    Since the ``execute`` function is differentiable, we can
+    also compute the gradient:
+
+    >>> qml.grad(cost_fn)(params, x)
+    (array([-0.0978434 , -0.19767681, -0.29552021]), array([5.37764278e-17]))
+
+    Finally, we can also compute any nth-order derivative. Let's compute the Jacobian
+    of the gradient (that is, the Hessian):
+
+    >>> x.requires_grad = False
+    >>> qml.jacobian(qml.grad(cost_fn))(params, x)
+    array([[-0.97517033,  0.01983384,  0.        ],
+           [ 0.01983384, -0.97517033,  0.        ],
+           [ 0.        ,  0.        , -0.95533649]])
+    """
+    # Default execution function; simply call device.batch execute
+    # and return no Jacobians.
+    execute_fn = lambda tapes: (device.batch_execute(tapes), [])
+
+    if gradient_fn == "device":
+        # gradient function is a device method
+
+        if accumulation == "forward":
+            # replace the forward execution function to return
+            # both results and gradients
+            execute_fn = device.execute_and_gradients
+
+        elif accumulation == "backward":
+            # replace the backward gradient computation
+            gradient_fn = device.gradients
+
+    if interface == "autograd":
+        return execute_autograd(tapes, device, execute_fn, gradient_fn)
+
+    raise ValueError(f"Unknown interface {interface}")

```

Comment:
So only `gradient_fn=="device"` is currently supported? What other options will we have?

Reply:
I don't know, is there anything else? Currently, all we support in master is:

- Gradient transforms (parameter shift, finite diff, CV param shift)
- Device gradients (adjoint, plugin implementations, API endpoints)

Only thing that falls outside of this is `backprop`, but backprop is unique in that this interface logic is not needed. No parameters are ever unwrapped, and the autodiff framework is in complete control of the entire process.

---

Note that missing from the above list is the reversible method. It should be implemented as a device-level method, but currently isn't. I think it might be a big undertaking to move it there.